library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,198 @@
+    /// the stack.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(thin_box)]
+    /// use std::boxed::ThinBox;
+    ///
+    /// let thin_slice = ThinBox::<[i32]>::new_unsize([1, 2, 3, 4]);
+    /// ```
+    pub fn new_unsize<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        let dyn_ref: &Dyn = &value;
+        let (_, meta) = (dyn_ref as *const Dyn).to_raw_parts();
+        let ptr = WithHeader::new(meta, value).expect("No allocation failure");
+        ThinBox { ptr, _marker: PhantomData }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized + Debug> Debug for ThinBox<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.deref())
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Deref for ThinBox<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts(value as *const (), metadata);
+        unsafe { &*pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> DerefMut for ThinBox<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts::<T>(value as *const (), metadata) as *mut T;
+        unsafe { &mut *pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Drop for ThinBox<T> {
+    fn drop(&mut self) {
+        unsafe {
+            let value = self.deref_mut();
+            let value = value as *mut T;
+            self.ptr.drop::<T>(value);
+        }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> ThinBox<T> {
+    fn meta(&self) -> <T as Pointee>::Metadata {
+        //  Safety:
+        //  -   NonNull and valid.
+        unsafe { *self.ptr.header() }
+    }
+
+    fn data(&self) -> *mut u8 {
+        self.ptr.value()
+    }
+}
+
+struct WithHeader<H>(*mut u8, PhantomData<H>);
+
+impl<H> WithHeader<H> {
+    fn new<T>(header: H, value: T) -> Option<WithHeader<H>> {
+        let layout = Self::alloc_layout(mem::size_of::<T>(), mem::align_of::<T>());
+        let aligned_header_size = Self::aligned_header_size(layout.align());
+
+        unsafe {
+            let ptr = NonNull::new(alloc::alloc(layout))?;
+
+            //  Safety:
+            //  -   The size is at least `aligned_header_size`.
+            let ptr = ptr.as_ptr().offset(aligned_header_size as isize) as *mut _;
+
+            let result = WithHeader(ptr, PhantomData);
+            ptr::write(result.header(), header);
+            ptr::write(result.value().cast(), value);
+
+            Some(result)
+        }
+    }
+
+    //  Safety:
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: *mut T) {
```

Comment:
This should definitely take a `&mut`

Reply:
It's probably fine because `&self` points at the `NonNull<u8>` while `value` points at (the neighborhood of) the pointee, so it's not actually overlapping, right?

Reply:
> Also, using the pointer here may violate an aliasing rule (not sure, but this signature seems dubious).

cc @RalfJung but my understanding was that between `&mut T` and `*mut T`, the latter was more flexible, and the former is the one that can cause more aliasing issues

Reply:
Yes, it seems a lot less dubious to use a pointer here.

Reply:
I swapped it back based on the consensus, marking this as resolved now.